scikits/cuda/cusparse.py

```diff
@@ -1,387 +1,1433 @@
+    """ multiply a sparse matrix, A, by dense matrix B:
+    C = alpha * transA(A) * B + beta * C.
 
+    higher level wrapper to cusparse<t>csrmm routines
     """
+    if check_inputs:
+        for item in [csrValA, csrRowPtrA, csrColIndA, B]:
+            if not isinstance(item, pycuda.gpuarray.GPUArray):
+                raise ValueError("csr*, B, must be pyCUDA gpuarrays")
+        if C is not None:
+            if not isinstance(C, pycuda.gpuarray.GPUArray):
+                raise ValueError("C must be a pyCUDA gpuarray or None")
+        # dense matrices must be in column-major order
+        if not B.flags.f_contiguous:
+            raise ValueError("Dense matrix B must be in column-major order")
+    if handle is None:
+        handle = misc._global_cusparse_handle
+
+    dtype = csrValA.dtype
+    if C is None:
+        if transA == CUSPARSE_OPERATION_NON_TRANSPOSE:
+            ldc = m
+        else:
+            ldc = k
+        alloc = misc._global_cusparse_allocator
+        C = gpuarray.zeros((ldc, n), dtype=dtype, order='F', allocator=alloc)
+    elif not C.flags.f_contiguous:
+        raise ValueError("Dense matrix C must be in column-major order")
+    if nnz is None:
+        nnz = csrValA.size
+    if ldb is None:
+        ldb = B.shape[0]
+    if ldc is None:
+        ldc = C.shape[0]
+
+    # perform some basic sanity checks
+    if check_inputs:
+        if csrValA.size != nnz:
+            raise ValueError("length of csrValA array must match nnz")
+        if (B.dtype != dtype) or (C.dtype != dtype):
+            raise ValueError("A, B, C must share a common dtype")
+        if ldb < B.shape[0]:
+            raise ValueError("ldb invalid for matrix B")
+        if ldc < C.shape[0]:
+            raise ValueError("ldc invalid for matrix C")
+        if (C.shape[1] != n) or (B.shape[1] != n):
+            raise ValueError("bad shape for B or C")
+        if transA == CUSPARSE_OPERATION_NON_TRANSPOSE:
+            if (ldb != k) or (ldc != m):
+                raise ValueError("size of A incompatible with B or C")
+        else:
+            if (ldb != m) or (ldc != k):
+                raise ValueError("size of A incompatible with B or C")
+        if csrRowPtrA.size != m+1:
+            raise ValueError("length of csrRowPtrA invalid")
+    if dtype == np.float32:
+        fn = cusparseScsrmm
+    elif dtype == np.float64:
+        fn = cusparseDcsrmm
+    elif dtype == np.complex64:
+        fn = cusparseCcsrmm
+    elif dtype == np.complex128:
+        fn = cusparseZcsrmm
+    else:
+        raise ValueError("unsupported sparse matrix dtype: %s" % dtype)
+    if toolkit_version >= (4, 1, 0):
+        fn(handle, transA, m, n, k, nnz, alpha, descrA, csrValA, csrRowPtrA,
+           csrColIndA, B, ldb, beta, C, ldc)
+    else:
+        fn(handle, transA, m, n, k, alpha, descrA, csrValA, csrRowPtrA,
+           csrColIndA, B, ldb, beta, C, ldc)
+    return C
+
+
+if toolkit_version >= (5, 5, 0):
```

Comment:
I agree decorators are probably a cleaner choice for skipping definition of the functions not present in a given version.  

I am not very experienced with decorators, but would there also be a way to use it in the case of the ```toolkit_version >= (4, 1, 0)``` code above?  I don't know how that would work or if it would be less clear in that case.  Here, the autogenerated cusparse<X>csrmm wrappers will expect a different number of arguments depending on the CUDA version  (i.e. an additional nnz argument was added starting in CUDA v4.1+ with the cusparse_v2.h headers).